parallel/client.py

```diff
@@ -0,0 +1,140 @@
+import numpy as np
+import time
+
+import pickle
+import grpc
+
+from parallel import service_pb2_grpc, service_pb2
+
+from concurrent import futures
+
+import os
+
+# (ytz): The classes in this file are designed to help provide a consistent API between
+# multiprocessing (typically for local cluster use) and gRPC (distributed and multi-node).
+
+class AbstractClient():
+
+    def submit(self, task_fn, *args):
+        """
+        Submit is an asynchronous method that will launch task_fn whose 
+        results will be collected a later point in time. The input task_fn
+        and its arguments should be picklable. See Python documentation for
+        pickle rules.
+
+        Parameters
+        ----------
+        task_fn: callable
+            A python function to be called
+
+        args: list
+            list of arguments for task_fn
+
+
+        Returns
+        -------
+        Future
+            A deferred object with a .result() method.
+
+        Usage:
+
+        client = ConcreteClient()
+
+        futures = []
+        for arg in args:
+            fut = client.submit(fn, *args)
+            futures.append(fut)
+
+        for fut in futures:
+            res = fut.result()
```

Comment:
The [examples in `tests/test_parallel.py`](https://github.com/proteneer/timemachine/blob/caa7a76311c17dafe22113543886486d1b284a8a/tests/test_parallel.py#L27-L34) look a bit more like this:

```suggestion
        futures = []
        for arg in args:
            fut = client.submit(task_fn, arg)
            futures.append(fut)
            
        res = []
        for fut in futures:
            res.append(fut.result())
```